libdrgn/arch_s390x.c

```diff
@@ -190,6 +190,66 @@ linux_kernel_get_initial_registers_s390x(const struct drgn_object *task_obj,
 	return err;
 }
 
+static struct drgn_error *
+get_initial_registers_from_struct_s390x(struct drgn_program *prog,
+					const void *buf, size_t size,
+					bool full_regset,
+					struct drgn_register_state **ret)
+{
+	if (size < (full_regset ? 208 : 144)) {
+		return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
+					 "registers are truncated");
+	}
+
+	struct drgn_register_state *regs;
+	if (full_regset)
+		regs = drgn_register_state_create(a15, true);
+	else
+		regs = drgn_register_state_create(pswa, true);
+
+	if (!regs)
+		return &drgn_enomem;
+
+	drgn_register_state_set_from_buffer(regs, pswm, (uint64_t *)buf);
+	drgn_register_state_set_from_buffer(regs, pswa, (uint64_t *)buf + 1);
+	drgn_register_state_set_range_from_buffer(regs, r0, r5, (uint64_t *)buf + 2);
+	drgn_register_state_set_range_from_buffer(regs, r6, r15, (uint64_t *)buf + 8);
+	if (full_regset)
+		drgn_register_state_set_range_from_buffer(regs, a0, a15, (uint64_t *)buf + 18);
+	drgn_register_state_set_pc_from_register(prog, regs, pswa);
+	*ret = regs;
+	return NULL;
+}
+
+static struct drgn_error *
+pt_regs_get_initial_registers_s390x(const struct drgn_object *obj,
+				    struct drgn_register_state **ret)
+{
+	return get_initial_registers_from_struct_s390x(drgn_object_program(obj),
```

Comment:
Indeed, i mixed up the structures. I changed that in my code, but it looks like i get an all zero buffer in pt_regs_get_initial_registers_s390x(). Haven't looked into the code why.

Reply:
just noted that the kernel module uses crash_setup_regs(), but s390 doesn't implement that.

Reply:
`Finished processing dependencies for drgn==0.0.21+69.gc1aa7d8fcd40
test_by_pid_dwarf (tests.linux_kernel.test_stack_trace.TestStackTrace) ... ok
test_by_pid_orc (tests.linux_kernel.test_stack_trace.TestStackTrace) ... skipped 's390x does not use ORC'
test_by_pt_regs (tests.linux_kernel.test_stack_trace.TestStackTrace) ... ok
test_by_task_struct (tests.linux_kernel.test_stack_trace.TestStackTrace) ... ok
test_local_variable (tests.linux_kernel.test_stack_trace.TestStackTrace) ... ok
test_locals (tests.linux_kernel.test_stack_trace.TestStackTrace) ... ok
test_prog (tests.linux_kernel.test_stack_trace.TestStackTrace) ... ok
test_registers (tests.linux_kernel.test_stack_trace.TestStackTrace) ... ok
test_sp (tests.linux_kernel.test_stack_trace.TestStackTrace) ... ok
test_stack__repr_pretty_ (tests.linux_kernel.test_stack_trace.TestStackTrace) ... ok

----------------------------------------------------------------------
Ran 10 tests in 0.682s

OK (skipped=1)`

Reply:
Awesome! Would it be possible to put your implementation of `crash_setup_regs()` in the test module itself, so that it will work without your patch? Something like:

```diff
diff --git a/tests/linux_kernel/kmod/drgn_test.c b/tests/linux_kernel/kmod/drgn_test.c
index 681de009..2db79b00 100644
--- a/tests/linux_kernel/kmod/drgn_test.c
+++ b/tests/linux_kernel/kmod/drgn_test.c
@@ -404,6 +404,20 @@ static struct task_struct *drgn_test_kthread;
 static DECLARE_COMPLETION(drgn_test_kthread_ready);
 struct pt_regs drgn_test_kthread_pt_regs;
 
+#if __s390x__
+// s390x doesn't implement crash_setup_regs().
+static inline void drgn_test_get_pt_regs(struct pt_regs *regs)
+{
+	// YOUR IMPLEMENTATION HERE
+}
+#else
+// crash_setup_regs() is an internal function for kexec, but it does exactly
+// what we need: saves the current registers in a struct pt_regs. If it goes
+// away or doesn't work on other architectures, we might need to do it ourselves
+// like we do for s390x.
+#define drgn_test_get_pt_regs(regs) crash_setup_regs(regs, NULL)
+#endif
+
  __attribute__((__optimize__("O0")))
 static void drgn_test_kthread_fn3(void)
 {
@@ -423,11 +437,7 @@ static void drgn_test_kthread_fn3(void)
 		}
 		if (kthread_should_park()) {
 			__set_current_state(TASK_RUNNING);
-
-			// This is an internal function for kexec, but I
-			// couldn't find a better way to get current pt_regs.
-			crash_setup_regs(&drgn_test_kthread_pt_regs, NULL);
-
+			drgn_test_get_pt_regs(&drgn_test_kthread_pt_regs);
 			kthread_parkme();
 			continue;
 		}
```

(I might end up doing that for every architecture anyways instead of relying on an implementation detail.)

Other than that, I think the only unresolved issue is what to do about the fallback unwinder in userspace. What do you think about my earlier suggestion:

> If the common case for userspace is no backchains, then fallback_unwind isn't very useful for it anyways, and we can make it always fail (i.e., return `&drgn_stop`) for userspace, relying solely on DWARF/`.eh_frame`.

Reply:
I copied your suggestion drgn_test_get_pt_regs(), test are fine with that on s390. For the 'userspace with no backchain' case i moved

```
f (!(prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL))
                return &drgn_stop;
```

to the beginning of fallback_unwind_s390x(), so it will never try to unwind userspace with the help of backchains.